orttraining/orttraining/test/training_api/common/synthetic_data_loader.h

```diff
@@ -74,24 +57,29 @@ struct SyntheticSampleBatch {
   void AddInt32Input(const std::vector<int64_t>& shape, int32_t low, int32_t high);
   void AddInt64Input(const std::vector<int64_t>& shape, int64_t low, int64_t high);
   void AddFloatInput(const std::vector<int64_t>& shape);
+  void AddBoolInput(const std::vector<int64_t>& shape);
 
   size_t NumOfInput() {
-    return data_vector_.size();
+    return input_count_;
   }
 
-  SyntheticInput* GetInputAtIndex(size_t index) {
-    return data_vector_[index].get();
+  SyntheticInput& GetInputAtIndex(size_t index) {
+    return data_vector_[index];
   }
 
  private:
-  std::vector<std::unique_ptr<SyntheticInput>> data_vector_;
+  template <typename T>
+  void AddIntInput(const std::vector<int64_t>& shape, T low, T high);
+
+  std::vector<SyntheticInput> data_vector_;
+  size_t input_count_{0};
 };
 
 struct SyntheticDataLoader {
   SyntheticDataLoader() {}
 
-  void AddSyntheticSampleBatch(std::unique_ptr<SyntheticSampleBatch> samples) {
-    sample_batch_collections_.emplace_back(std::move(samples));
+  void AddSyntheticSampleBatch(SyntheticSampleBatch& samples) {
```

Comment:
nit: could take `SyntheticSampleBatch&& samples` parameter and then `sample_batch_collections_.emplace_back(std::move(sample));`

Reply:
emplace() is not needed. Simply push_back(std::move())

Reply:
OK. Sorry I did not understand how could we pass a lvalue (e.g. the samples) as a rvalue reference (e.g. SyntheticSampleBatch&&)?  The compiler Compians "/bert_ort/pengwa/onnxruntime/orttraining/orttraining/test/training_api/trainer/trainer.cc:161:43: error: cannot bind rvalue reference of type ‘onnxruntime::training::test::training_api::SyntheticSampleBatch&&’ to lvalue of type ‘onnxruntime::training::test::training_api::SyntheticSampleBatch’
  161 |       data_loader.AddSyntheticSampleBatch(sample);"

Reply:
at the call site, you can do something like:
```c++
data_loader.AddSyntheticSampleBatch(std::move(sample));
```